File: backend/models/reportes/Alertas.py
# backend/models/reportes/Alertas.py
from __future__ import annotations

from datetime import datetime, date
from typing import Optional, List

from sqlalchemy import Column, Integer, String, Date, DateTime, ForeignKey, Boolean
from sqlalchemy.orm import Session
from pydantic import BaseModel

from backend.database import Base


class AlertaModel(Base):
    __tablename__ = "reportes_alertas"

    id = Column(Integer, primary_key=True, index=True)
    empresa_id = Column(Integer, ForeignKey("empresas.id"), nullable=False)
    granja_id = Column(Integer, ForeignKey("granjas.id"), nullable=False)

    fecha = Column(Date, nullable=False)
    tipo = Column(String(50), nullable=False)  # Sanidad, Productividad, Bioseguridad, Costos, Infraestructura, Otro
    nivel = Column(String(50), nullable=False)  # Crítico, Precaución, Informativo
    descripcion = Column(String(500), nullable=False)
    responsable = Column(String(120), nullable=False)
    cerrado = Column(Boolean, nullable=False, default=False)
    acciones = Column(String(500), nullable=True)

    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(
        DateTime,
        nullable=False,
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
    )


class AlertaBase(BaseModel):
    empresa_id: int
    granja_id: int
    fecha: date
    tipo: str  # Sanidad, Productividad, Bioseguridad, Costos, Infraestructura, Otro
    nivel: str  # Crítico, Precaución, Informativo
    descripcion: str
    responsable: str
    cerrado: bool = False
    acciones: Optional[str] = None


class AlertaCreate(AlertaBase):
    pass


class AlertaUpdate(BaseModel):
    tipo: Optional[str] = None
    nivel: Optional[str] = None
    descripcion: Optional[str] = None
    responsable: Optional[str] = None
    cerrado: Optional[bool] = None
    acciones: Optional[str] = None


class AlertaRead(AlertaBase):
    id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class AlertaRepository:
    def __init__(self, db: Session):
        self.db = db

    def listar_por_granja(
        self,
        empresa_id: int,
        granja_id: int,
        tipo: Optional[str] = None,
        solo_abiertas: bool = False,
    ) -> List[AlertaModel]:
        q = (
            self.db.query(AlertaModel)
            .filter(
                AlertaModel.empresa_id == empresa_id,
                AlertaModel.granja_id == granja_id,
            )
        )
        if tipo:
            q = q.filter(AlertaModel.tipo == tipo)
        if solo_abiertas:
            q = q.filter(AlertaModel.cerrado == False)
        return (
            q.order_by(
                AlertaModel.fecha.desc(),
                AlertaModel.id.desc(),
            )
            .all()
        )

    def crear(self, data: AlertaCreate) -> AlertaModel:
        reg = AlertaModel(
            empresa_id=data.empresa_id,
            granja_id=data.granja_id,
            fecha=data.fecha,
            tipo=data.tipo,
            nivel=data.nivel,
            descripcion=data.descripcion,
            responsable=data.responsable,
            cerrado=data.cerrado,
            acciones=data.acciones,
        )
        self.db.add(reg)
        self.db.commit()
        self.db.refresh(reg)
        return reg

    def actualizar(self, alerta_id: int, data: AlertaUpdate) -> AlertaModel:
        reg = self.db.query(AlertaModel).filter(AlertaModel.id == alerta_id).first()
        if not reg:
            return None
        for key, value in data.dict(exclude_unset=True).items():
            setattr(reg, key, value)
        self.db.commit()
        self.db.refresh(reg)
        return reg

    def obtener_por_id(self, alerta_id: int) -> AlertaModel:
        return self.db.query(AlertaModel).filter(AlertaModel.id == alerta_id).first()

File: frontend/src/services/reportes/Alertas.ts
// frontend/src/services/reportes/Alertas.ts
import API_BASE from "../../config/api";

export type Alerta = {
  id: number;
  empresa_id: number;
  granja_id: number;
  fecha: string;
  tipo: "Sanidad" | "Productividad" | "Bioseguridad" | "Costos" | "Infraestructura" | "Otro";
  nivel: "Crítico" | "Precaución" | "Informativo";
  descripcion: string;
  responsable: string;
  cerrado: boolean;
  acciones: string | null;
  created_at: string;
  updated_at: string;
};

export type NuevaAlerta = Omit<
  Alerta,
  "id" | "created_at" | "updated_at"
>;

export type ActualizarAlerta = Partial<NuevaAlerta>;

const EMPRESA_ID = 1;
const GRANJA_ID = 1;

function mapApiToAlerta(api: any): Alerta {
  return {
    id: api.id,
    empresa_id: api.empresa_id,
    granja_id: api.granja_id,
    fecha: api.fecha,
    tipo: api.tipo,
    nivel: api.nivel,
    descripcion: api.descripcion,
    responsable: api.responsable,
    cerrado: api.cerrado,
    acciones: api.acciones ?? null,
    created_at: api.created_at,
    updated_at: api.updated_at,
  };
}

export async function getAlertas(
  tipo?: string,
  soloAbiertas?: boolean
): Promise<Alerta[]> {
  const url = new URL(
    `${API_BASE}/reportes/alertas/`,
    window.location.origin
  );
  url.searchParams.set("empresa_id", String(EMPRESA_ID));
  url.searchParams.set("granja_id", String(GRANJA_ID));
  if (tipo) {
    url.searchParams.set("tipo", tipo);
  }
  if (soloAbiertas) {
    url.searchParams.set("solo_abiertas", "true");
  }

  const res = await fetch(url.toString());
  if (!res.ok) {
    throw new Error("Error al obtener alertas");
  }
  const data = await res.json();
  return data.map(mapApiToAlerta);
}

export async function addAlerta(
  payload: Omit<NuevaAlerta, "empresa_id" | "granja_id">
): Promise<Alerta> {
  const body = {
    empresa_id: EMPRESA_ID,
    granja_id: GRANJA_ID,
    ...payload,
  };
  const res = await fetch(`${API_BASE}/reportes/alertas/`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!res.ok) {
    throw new Error("Error al registrar alerta");
  }
  const data = await res.json();
  return mapApiToAlerta(data);
}

export async function updateAlerta(
  alertaId: number,
  payload: ActualizarAlerta
): Promise<Alerta> {
  const res = await fetch(`${API_BASE}/reportes/alertas/${alertaId}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    throw new Error("Error al actualizar alerta");
  }
  const data = await res.json();
  return mapApiToAlerta(data);
}

export async function getAlertaById(alertaId: number): Promise<Alerta> {
  const res = await fetch(`${API_BASE}/reportes/alertas/${alertaId}`);
  if (!res.ok) {
    throw new Error("Error al obtener alerta");
  }
  const data = await res.json();
  return mapApiToAlerta(data);
}
